classes/search/cmsfield.php

```diff
@@ -0,0 +1,187 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Define search area.
+ *
+ * @package    mod_cms
+ * @author     Tomo Tsuyuki <tomotsuyuki@catalyst-au.com>
+ * @copyright  2024 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace mod_cms\search;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/mod/cms/lib.php');
+
+/**
+ * Define search area.
+ *
+ * @package    mod_cms
+ * @author     Tomo Tsuyuki <tomotsuyuki@catalyst-au.com>
+ * @copyright  2024 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cmsfield extends \core_search\base_mod {
+
+    /**
+     * @var array Internal quick static cache.
+     */
+    protected $cmsdata = [];
+
+    /**
+     * Returns recordset containing required data for indexing cmsfield records.
+     *
+     * @param int $modifiedfrom timestamp
+     * @param \context|null $context Optional context to restrict scope of returned results
+     * @return moodle_recordset|null Recordset (or null if no results)
+     */
+    public function get_document_recordset($modifiedfrom = 0, \context $context = null) {
+        global $DB;
+
+        list ($contextjoin, $contextparams) = $this->get_context_restriction_sql(
+                $context, 'cms', 'mc');
+        if ($contextjoin === null) {
+            return null;
+        }
+
+        $sql = "SELECT mcd.*, mc.id AS cmsid, mc.course AS courseid, mcf.name AS fieldname
+                  FROM {customfield_data} mcd
+                  JOIN {cms} mc ON mc.id = mcd.instanceid
+                  JOIN {customfield_field} mcf ON mcf.id = mcd.fieldid
+                  JOIN {customfield_category} mcc ON mcf.categoryid = mcc.id
+          $contextjoin
+                 WHERE mcd.timemodified >= ? AND mcc.component = 'mod_cms' AND mcc.area = 'cmsfield'
```

Comment:
Do we also need to filter by supported field types?

Reply:
Type added.